crates/ln-dlc-node/src/ln/probes.rs

```diff
@@ -0,0 +1,69 @@
+use anyhow::Result;
+use lightning::ln::channelmanager::PaymentId;
+use lightning::routing::router::Path;
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+use std::time::Duration;
+use tokio::sync::RwLock;
+use tokio::time::error::Elapsed;
+
+#[derive(Clone, Debug)]
+pub enum ProbeStatus {
+    Succeeded { path: Path },
+    Failed,
+}
+
+#[derive(Default)]
+pub struct Probes {
+    inner: RwLock<HashMap<PaymentId, Option<ProbeStatus>>>,
+}
+
+impl Probes {
+    pub async fn register(&self, payment_id: PaymentId) {
+        self.inner.write().await.insert(payment_id, None);
+    }
+
+    pub async fn update_status(&self, payment_id: PaymentId, new_status: ProbeStatus) {
+        match self.inner.write().await.entry(payment_id) {
+            Entry::Occupied(mut entry) => {
+                let old_status = entry.insert(Some(new_status.clone()));
+
+                tracing::debug!(%payment_id, ?old_status, ?new_status, "Updated probe");
+            }
+            Entry::Vacant(entry) => {
+                entry.insert(Some(new_status.clone()));
+
+                tracing::debug!(%payment_id, ?new_status, "Updating unregistered probe");
+            }
+        }
+    }
+
+    pub async fn wait_until_finished(
```

Comment:
🔧 Instead of polling here couldn't we simply subscribe to the `Event::ProbeSuccessful` and `Event::ProbeFailed` events?

Reply:
I think that would also allow us to get rid of the `RwLock`

Reply:
> Instead of polling here couldn't we simply subscribe to the Event::ProbeSuccessful and Event::ProbeFailed events?

We could and I tried that first. But this was a lot easier to implement _for me_, so I went this way instead.

If you're really keen on that idea I'm happy to review a PR that goes in that direction.